packages/ws/src/ws/WebSocketShard.ts

```diff
@@ -0,0 +1,515 @@
+	[WebSocketShardEvents.Resumed]: [];
+};
+
+export interface WebSocketShardDestroyOptions {
+	reason?: string;
+	code?: number;
+	recover?: WebSocketShardDestroyRecovery;
+}
+
+export class WebSocketShard extends AsyncEventEmitter<WebSocketShardEventsMap> {
+	private connection: WebSocket | null = null;
+
+	private readonly id: number;
+
+	private useIdentifyCompress = false;
+
+	private inflate: Inflate | null = null;
+	private readonly textDecoder = new TextDecoder();
+
+	private status: WebSocketShardStatus = WebSocketShardStatus.Idle;
+
+	private replayedEvents = 0;
+
+	private isAck = true;
+
+	private sendRateLimitState = {
+		remaining: 120,
+		resetAt: Date.now(),
+	};
+
+	private heartbeatInterval: NodeJS.Timer | null = null;
+	private lastHeartbeatAt = -1;
+
+	private session: SessionInfo | null = null;
+
+	private readonly sendQueue = new AsyncQueue();
+
+	public readonly strategy: IContextFetchingStrategy;
+
+	public constructor(strategy: IContextFetchingStrategy, id: number) {
+		super();
+		this.strategy = strategy;
+		this.id = id;
+	}
+
+	public async connect() {
+		if (this.status !== WebSocketShardStatus.Idle) {
+			throw new Error("Tried to connect a shard that wasn't idle");
+		}
+
+		const data = this.strategy.options.gatewayInformation;
+
+		const { version, encoding, compression } = this.strategy.options;
+		const params = new URLSearchParams({ v: version, encoding });
+		if (compression) {
+			const zlib = await getZlibSync();
+			if (zlib) {
+				params.append('compress', compression);
+				this.inflate = new zlib.Inflate({
+					chunkSize: 65535,
+					to: 'string',
+				});
+			} else if (!this.useIdentifyCompress) {
+				this.useIdentifyCompress = true;
+				console.warn(
+					'WebSocketShard: Compression is enabled but zlib-sync is not installed, falling back to identify compress',
+				);
+			}
+		}
+
+		const url = `${data.url}?${params.toString()}`;
+		this.debug([`Connecting to ${url}`]);
+		const connection = new WebSocket(url, { handshakeTimeout: this.strategy.options.handshakeTimeout ?? undefined })
+			/* eslint-disable @typescript-eslint/no-misused-promises */
+			.on('message', this.onMessage.bind(this))
+			.on('error', this.onError.bind(this))
+			.on('close', this.onClose.bind(this));
+		/* eslint-enable @typescript-eslint/no-misused-promises */
+
+		connection.binaryType = 'arraybuffer';
+		this.connection = connection;
+
+		this.status = WebSocketShardStatus.Connecting;
+
+		await this.waitForEvent(WebSocketShardEvents.Hello, this.strategy.options.helloTimeout);
+
+		const session = this.session ?? (await this.strategy.retrieveSessionInfo(this.id));
+		if (session?.shardCount === this.strategy.options.shardCount) {
+			await this.resume(session);
+		} else {
+			await this.identify();
+		}
+	}
+
+	public async destroy(options: WebSocketShardDestroyOptions = {}) {
+		if (this.status === WebSocketShardStatus.Idle) {
+			throw new Error('Tried to destroy a shard that was idle');
```

Comment:
I'd just make this no-op instead of throw

Reply:
Will add a debug log